plugin/checkout_handlers.go

```diff
@@ -0,0 +1,118 @@
+}
+
+// StorageHandler's sole responsibility is to communicate with storage regarding check-outs.
+type StorageHandler struct{}
+
+// CheckOut will return:
+//  - Nil if it was successfully able to perform the requested check out.
+//  - ErrCurrentlyCheckedOut if the account was already checked out.
+//  - Some other err if it was unable to complete successfully.
+func (h *StorageHandler) CheckOut(ctx context.Context, storage logical.Storage, serviceAccountName string, checkOut *CheckOut) error {
+	if err := validateInputs(ctx, storage, serviceAccountName, checkOut, true); err != nil {
+		return err
+	}
+	// Check if the service account is currently checked out.
+	if entry, err := storage.Get(ctx, checkoutStoragePrefix+serviceAccountName); err != nil {
+		return err
+	} else if entry != nil {
+		return ErrCurrentlyCheckedOut
+	}
+	// Since it's not, store the new check-out.
+	entry, err := logical.StorageEntryJSON(checkoutStoragePrefix+serviceAccountName, checkOut)
+	if err != nil {
+		return err
+	}
+	return storage.Put(ctx, entry)
+}
+
+// CheckIn will return nil error if it was able to successfully check in an account.
+// If the account was already checked in, it still returns no error.
+func (h *StorageHandler) CheckIn(ctx context.Context, storage logical.Storage, serviceAccountName string) error {
+	if err := validateInputs(ctx, storage, serviceAccountName, nil, false); err != nil {
+		return err
+	}
+	// We simply take checkouts out of storage when they're checked in.
+	return storage.Delete(ctx, checkoutStoragePrefix+serviceAccountName)
+}
+
+// Status returns either:
+//  - A *CheckOut and nil error if the serviceAccountName is currently checked out.
+//  - A nil *CheckOut and nil error if the serviceAccountName is not currently checked out.
+//  - A nil *CheckOut and populated err if the state cannot be determined.
+func (h *StorageHandler) Status(ctx context.Context, storage logical.Storage, serviceAccountName string) (*CheckOut, error) {
+	if err := validateInputs(ctx, storage, serviceAccountName, nil, false); err != nil {
+		return nil, err
+	}
+	entry, err := storage.Get(ctx, checkoutStoragePrefix+serviceAccountName)
+	if err != nil {
+		return nil, err
+	}
+	if entry == nil {
+		return nil, nil
+	}
+	checkOut := &CheckOut{}
+	if err := entry.DecodeJSON(checkOut); err != nil {
+		return nil, err
+	}
+	return checkOut, nil
+}
+
+// validateInputs is a helper function for ensuring that a caller has satisfied all required arguments.
+func validateInputs(ctx context.Context, storage logical.Storage, serviceAccountName string, checkOut *CheckOut, checkOutRequired bool) error {
+	if ctx == nil {
+		return errors.New("ctx is required")
+	}
+	if storage == nil {
+		return errors.New("storage is required")
+	}
+	if serviceAccountName == "" {
+		return errors.New("serviceAccountName is required")
+	}
+	if checkOutRequired && checkOut == nil {
+		return errors.New("checkOut is required")
+	}
+	return nil
```

Comment:
Should the above error messages be constants like `ErrCurrentlyCheckedOut`? It's not explicitly stated as an idiom in [Effective Go](https://golang.org/doc/effective_go.html) but it's a pattern shown a few times in that document, might be a good idea

Reply:
At the present time, the errors are just strings that will be returned, raw, to the user in a logical error response. Since I will never need to test their type, it seems like strongly typed errors aren't needed at this point.

Reply:
That makes sense, thanks 👍